pkg/storegateway/bucket.go

```diff
@@ -1043,6 +950,146 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 	return err
 }
 
+// synchronousSeriesSet returns seriesSet that contains the requested series. It returns a cleanup func. The cleanup func
+// should be invoked always when non-nil; even when the returned error is non-nil. synchronousSeriesSet must be called
+// while holding s.mtx.RLock(). synchronousSeriesSet will call s.mtx.RUnlock regardless of its return values
```

Comment:
From a design perspective, I don't like the idea that we have to call this function holding `s.mtx.RLock()` but then this function can unlock/relock it. I would like to remove the requirement that this function must be called with `s.mtx.RLock()` held and let this function managing the lock itself.

I left some comments to show how I think we can do it.